### PR TITLE
PR: Update Spyder installer base environment with conda-lock file

### PR DESCRIPTION
**Pull request by mrclary**
## Description of Changes

* CI now builds a conda-lock file for the installer base environment in addition to the runtime environment.
* UpdateManager _updates_ the base environment and _replaces_ the runtime environment using conda-lock files.
* The installer base environment pins all explicit dependencies, including Python.

I've tested the update process for macOS, Linux, and Windows on my local machine and VMs. Both the base and runtime environments appear to update correctly. This is accomplished by installing 6.0.0b3 using the installer, then patching Spyder source code in the runtime environment with changes made in this PR. Additionally I spoof the Spyder version to 6.0.0b2 and change the github url to my fork. base and runtime lock files are created locally and uploaded to a 6.0.0b3.post1 releasse on my fork.

Note that using a lock file for the base environment precludes excluding packages, so `pip` cannot be excluded.

Note that updating the Python minor version of the base environment results in a broken base environment; only micro versions can be updated with the conda-lock file. When updating the minor version of Python in the base environment, a new installer must be downloaded, therefore updates of Python minor versions to the base environment should be reserved for Spyder major updates.

Because the runtime environment is removed and re-built with the lock file, Python can be updated in this environment with minor/micro Spyder updates and does not require using the installer.

### CODE REVIEW

File: installers-conda/build-environment.yml [hunk 1]
```diff
@@ -2,13 +2,13 @@ name: spy-inst
...
-  - constructor >=3.6.0
+  - conda =24.5.0
+  - conda-build =24.5.1
+  - conda-lock =2.5.7
+  - conda-standalone =24.5.0
+  - constructor =3.8.1
   - gitpython
   - mamba
-  - menuinst >=2.1.0
+  - menuinst =2.1.1
```
Comment (ccordoba12) [inline installers-conda/build-environment.yml:12]: I think it'd be a good idea to add a simple and dummy dependency here to check that the installers base env is updated between rc1 and rc2.

I know we're waiting for a new menuinst version to be released but since there's no clarity about it, I think that could be a good test for this PR.

File: spyder/plugins/updatemanager/scripts/install.bat [hunk 2]
```diff
@@ -24,20 +26,49 @@ echo.
...
-IF not "%install_file%"=="" (
+IF exist "%install_file%" (
     call :install_subroutine
     goto exit
 )
 
 :exit
 exit %ERRORLEVEL%
 
+:wait_for_spyder_quit
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/scripts/install.bat:44]: How can we check that these changes to the update script are working as expected?

File: spyder/plugins/updatemanager/scripts/install.sh [hunk 3]
```diff
@@ -2,18 +2,22 @@
 
 unset HISTFILE  # Do not write to history with interactive shell
 
-while getopts "i:c:p:" option; do
+while getopts "i:c:p:b:r:" option; do
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/scripts/install.sh:5]: And these?

File: spyder/plugins/updatemanager/workers.py [hunk 4]
```diff
@@ -229,15 +229,18 @@ class WorkerDownloadInstaller(BaseWorker):
...
-        self.installer_path = installer_path
-        self.installer_size_path = installer_size_path
+        self.installer_paths = installer_paths
         self.error = None
         self.cancelled = False
         self.paused = False
 
+        self.installer_dir = osp.dirname(
+            next(iter(installer_paths.values()))['file']
+        )
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/workers.py:242]: I don't understand what this code does. Could you rewrite it in a more standard way? Or at least could you write a comment about it to explain it? Thanks!

**Review by ccordoba12 — COMMENTED** [review]
Thanks @mrclary for your work on this!

**Comment by ccordoba12** [comment]
@mrclary, one more question for you: is it possible to zip all lock files into a single file and upload only that artifact  to the GH release?

I say that because Github displays all artifacts for the latest release, and I think having too many lock files would be distracting for users.

File: installers-conda/build-environment.yml (hunk 1 above)
Comment (mrclary) [inline installers-conda/build-environment.yml:12]: There are not significant changes from 2.1.0 to 2.1.1, so we can pin 2.1.0 for rc1 and then pin to 2.1.1 for rc2 and check that the base environment was updated.

```suggestion
  - menuinst =2.1.0
```

File: spyder/plugins/updatemanager/scripts/install.bat (hunk 2 above)
Comment (mrclary) [reply spyder/plugins/updatemanager/scripts/install.bat:44]: I've checked the update process for all three platforms using my local mac and Windows and Linux VMs. I did this by installing 6.0.0b3 using the installer, then patching the runtime environment with the changes in this PR and also spoofing the Spyder version to 6.0.0b2 and also changing the Github url to point to my fork. My fork has the runtime and base environment lock files in the 6.0.0b3 release.

Is this what you meant?

**Comment by mrclary** [comment]
> @mrclary, one more question for you: is it possible to zip all lock files into a single file and upload only that artifact to the GH release?

Perhaps this can be done...

If the installer jobs only upload the artifacts (but not to the release), perhaps a separate job after the installer jobs can retrieve all the artifacts, zip all the lock files, and upload all the artifacts to the release. This would require minor changes to the installer workflow, and some changes to the update manager.

I'll look into it.

File: spyder/plugins/updatemanager/scripts/install.bat (hunk 2 above)
Comment (ccordoba12) [reply spyder/plugins/updatemanager/scripts/install.bat:42]: Yeah, it is. Thanks for doing the extra step of manually checking that the update process is working as expected with the changes you've done here.

File: spyder/plugins/updatemanager/scripts/install.bat [hunk 5]
```diff
@@ -50,10 +51,14 @@ exit %ERRORLEVEL%
     goto :EOF
 
 :update_subroutine
-    echo Updating Spyder
-
-    %conda% update -p %prefix% -y --file %install_file%
-    set /P =Press return to exit...
+    for %%C in ("%install_file%") do set installer_dir=%%~dpC
+    pushd %installer_dir%
+    tar -xf %install_file%
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/scripts/install.bat:56]: Is `tar` available by default on Windows? Or does this executable come from a conda package?

File: spyder/plugins/updatemanager/scripts/install.sh [hunk 6]
```diff
@@ -12,8 +12,19 @@ done
 shift $(($OPTIND - 1))
 
 update_spyder(){
-    $conda update -p $prefix -y --file $install_file
-    read -p "Press return to exit..."
+    # Unzip installer file
+    pushd $(dirname $install_file)
+    unzip -o $install_file
```
Comment (ccordoba12) [inline spyder/plugins/updatemanager/scripts/install.sh:17]: Also, does this `unzip` come from a conda package?

**Review by ccordoba12 — COMMENTED** [review]
A couple of minor questions above the commands used to unzip the new file that contains all lock files.

The rest looks good to me.

File: spyder/plugins/updatemanager/scripts/install.bat (hunk 5 above)
Comment (mrclary) [reply spyder/plugins/updatemanager/scripts/install.bat:56]: `tar` is available be default on Windows 10 or greater. Do we need to support older Windows versions?

File: spyder/plugins/updatemanager/scripts/install.sh (hunk 6 above)
Comment (mrclary) [reply spyder/plugins/updatemanager/scripts/install.sh:17]: `unzip` is available by default on both macOS and Linux.

Comment (mrclary) [reply spyder/plugins/updatemanager/scripts/install.sh:17]: it is not a conda package

**Comment by mrclary** [comment]
Okay, the osx-arm64 lock file now has `asyncssh=2.14.1` and I've successfully tested the update process on all 4 platforms.

**Review by ccordoba12 — APPROVED** [review]
Looks good to me now, thanks @mrclary!

**Comment by ccordoba12** [comment]
About your notes:

> Note that using a lock file for the base environment precludes excluding packages, so pip cannot be excluded.

Ok, but is it possible for users to select the base env either in `Preferences > Main interpretrer` or through the `Consoles > New console in environment` menu?

If so, we need to prevent users to do that before 6.0 so that they can't install packages in that env,

> Note that updating the Python minor version of the base environment results in a broken base environment; only micro versions can be updated with the conda-lock file. When updating the minor version of Python in the base environment, a new installer must be downloaded, therefore updates of Python minor versions should be reserved for Spyder major updates.

I think that's too restrictive. Instead, we could update Python every two minor Spyder versions (i.e. we should update to Python 3.12 in Spyder 6.2). But we can talk about that in another issue.

**Comment by mrclary** [comment]
> > Note that using a lock file for the base environment precludes excluding packages, so pip cannot be excluded.
> 
> Ok, but is it possible for users to select the base env either in `Preferences > Main interpretrer` or through the `Consoles > New console in environment` menu?

No, users cannot select the base environment through either of these methods. The installer explicitly prevents the base and runtime environments from being registered, so they don't show up in `conda info -e` and Spyder does not list them in the possible interpreters. The runtime environment is only available as "Same as Spyder" interpreter selection.

**Comment by mrclary** [comment]
> > Note that updating the Python minor version of the base environment results in a broken base environment; only micro versions can be updated with the conda-lock file. When updating the minor version of Python in the base environment, a new installer must be downloaded, therefore updates of Python minor versions should be reserved for Spyder major updates.
> 
> I think that's too restrictive. Instead, we could update Python every two minor Spyder versions (i.e. we should update to Python 3.12 in Spyder 6.2). But we can talk about that in another issue.

So while investigating this, I've discovered another issue. When updating the runtime environment with the lock file, obsolete packages are not removed. Updated packages are correctly replaced and new packages are added, but if a package in the old environment is no longer needed, it is not removed. This is not good.

I think the best solution is to remove the runtime environment and recreate it with the lock file, rather than try to update it. This will also solve the issue of updating Python in the runtime environment, i.e. we could update Spyder to use Python 3.12 at any time.

However, there could still be a potential issue updating the base environment. Here, we cannot remove and recreate the environment. The base environment does not need to have the same Python version as Spyder's runtime, so we could still refrain from updating Python in the base environment until major updates to Spyder. So the only issue is whether the base environment update will ever require _removing_ unnecessary packages. This may be an extremely low risk since there isn't an extensive list of packages in the base environment (only Python, conda, mamba, and menuinst are specified).

What are your thoughts, @ccordoba12?

**Comment by ccordoba12** [comment]
> So while investigating this, I've discovered another issue. When updating the runtime environment with the lock file, obsolete packages are not removed.
> I think the best solution is to remove the runtime environment and recreate it with the lock file, rather than try to update it. This will also solve the issue of updating Python in the runtime environment, i.e. we could update Spyder to use Python 3.12 at any time.

I saw that you already implemented this, but I kind of disagree with doing it for every release. That's because it defeats the purpose of the update process, which is to have a simple and fast update. The fact that obsolete packages are not removed is (in my opinion) a minor nuisance that we can deal with every minor (or two minor releases).

In other words, I wouldn't like users to download close to 400 mb of packages every time a new version is available. That's almost the same as downloading the full installer.

**Comment by mrclary** [comment]
### TLDR:
Suggested changes:
* Let's _update_ the runtime environment for micro updates and _replace_ the runtime environment on minor updates. Installer is still required for major updates.
* Let's retain the package cache in the base environment in order to minimize required downloads in subsequent updates. We should still run a `conda clean --packages` in the post install script to save disk space.

### Details
> I saw that you already implemented this,

I can revert, but wanted to test an implementation in order to have it ready as soon as possible should we want it.

> but I kind of disagree with doing it for every release. That's because it defeats the purpose of the update process, which is to have a simple and fast update. The fact that obsolete packages are not removed is (in my opinion) a minor nuisance that we can deal with every minor (or two minor releases).

I agree that this is not ideal. Replacing the runtime environment is the only way to update the Python version outside of a major release. Perhaps the update manager can also determine whether the update is a minor update or a micro (bug fix) update? Then it could instruct the install script to either update the runtime environment for micro updates or remove and re-create the environment for minor updates and we could ensure that Python is only updated on minor updates. The update manager will still use a new installer for major updates.

> In other words, I wouldn't like users to download close to 400 mb of packages every time a new version is available. That's almost the same as downloading the full installer.

Conda will download every package listed in the lock file _if_ it does not already exist in the `pkgs` cache directory of the base environment. Currently, we have the installer clear the `pkgs` directory after the install is complete in order to save disk space, so the first minor or micro update will download all the packages anyway, regardless whether we update the runtime environment or replace it. Subsequent updates will only download missing packages. We could instruct the installer to leave the `pkgs` directory so even the first update will download fewer packages.

For what it's worth, in my testing this PR, downloading and installing with the installer takes much longer than the install script to download all packages and _replace_ the environment, though the latter does take longer than just an update process.

**Comment by ccordoba12** [comment]
Since @dalthviz is mostly unavailable this week, I'm going to merge this one because we need it for rc1.

Thanks @again mrclary for your hard work on our installers!
